frontend/src/core/hooks/useStream.js

```diff
@@ -1,61 +1,284 @@
+      newBoundary.include_start || extensionBoundary.include_start;
 
-const useJournalEntries = ({
-  searchQuery,
-  start_time,
-  end_time,
-  include_start,
-  include_end,
-  updateStreamBoundaryWith,
-}) => {
-  // set our get method
-  const getStream =
-    (searchTerm, start_time, end_time, include_start, include_end) =>
-    async () => {
-      // Request with params to streams
-      const response = await StreamService.getStream({
-        searchTerm: searchTerm,
-        start_time: start_time,
-        end_time: end_time,
-        include_start: include_start,
-        include_end: include_end,
-      });
-
-      // new events from stream
-      const newEventsList = response.data.stream.map((event) => ({
-        ...event,
-      }));
-
-      return {
-        data: [...newEventsList],
-        boundaries: { ...response.data.boundaries, update: false },
-      };
-    };
-
-  const { data, isLoading, refetch, isFetching, remove } = useQuery(
-    ["stream", searchQuery, start_time, end_time],
-    getStream(searchQuery, start_time, end_time, include_start, include_end),
+    if (
+      !newBoundary.end_time ||
+      (extensionBoundary.end_time &&
+        extensionBoundary.end_time >= newBoundary.end_time)
+    ) {
+      newBoundary.end_time = extensionBoundary.end_time;
+      newBoundary.include_end =
+        newBoundary.include_end || extensionBoundary.include_end;
+    }
+
+    newBoundary.include_end =
+      newBoundary.include_end || extensionBoundary.include_end;
+
+    setStreamBoundary(newBoundary);
+    return newBoundary;
+  };
+
+  const getEvents = async (customStreamBoundary) => {
+    let requestStreamBoundary = customStreamBoundary;
+    if (!requestStreamBoundary) {
+      requestStreamBoundary = streamBoundary;
+    }
+    const response = await StreamService.getEvents({
+      q: streamQuery,
+      ...requestStreamBoundary,
+    });
+    return response.data;
+  };
+
+  const {
+    isLoading: eventsIsLoading,
+    refetch: eventsRefetch,
+    isFetching: eventsIsFetching,
+    remove: eventsRemove,
+  } = useQuery(
+    ["stream-events", streamQuery],
+    () => {
+      if (isStreamBoundaryEmpty()) {
+        return null;
+      }
+      return getEvents();
+    },
     {
-      //refetchInterval: refreshRate,
       ...queryCacheProps,
       keepPreviousData: true,
-      retry: 3,
-      onSuccess: (response) => {
-        // response is object which return condition in getStream
-        // TODO(andrey): Response should send page parameters inside "boundary" object (can be null).
-        updateStreamBoundaryWith(response.boundaries);
+      retry: 2,
+      onSuccess: (newEvents) => {
+        if (newEvents && newEvents.stream_boundary && newEvents.events) {
+          setEvents([...newEvents.events]);
+          updateStreamBoundaryWith(newEvents.stream_boundary);
+        }
       },
     }
   );
 
+  const { refetch: loadOlderEvents, isFetching: loadOlderEventsIsFetching } =
```

Comment:
@zomglings I think you should not duplicate query in line 138

Reply:
Instead I think would be cleaner to define 

{refetch} = useQuery(....)

const getOlder() = () => refetch({ direction: `older` }) 
const getOlder() = () => refetch({ direction: `newer` })

Reply:
👍 will work on this today.

Reply:
Experimented with this and decided to keep things as they are. I might be missing something about `react-query` and if so I am happy to make the change in a future PR.

What I discovered. According to the [`useQuery` API](https://react-query.tanstack.com/reference/useQuery):
1. `refetch` always has the signature: `(options: {throwOnError: boolean, cancelRefetch: boolean}) => Promise<UseQueryResult>`. This means that it is not possible to pass arguments to `getEvents` using `refetch`.
2. The `queryFn` passed to `useQuery` (which ultimately calls `getEvents`) must have the signature: `(context: QueryFunctionContext) => Promise<TData>`. `QueryFunctionContext` does not give an easy way to pass arguments to the `getEvents` call that the `queryFn` must make.

This means that the only option left is to define `getEvents` as:
```js
const getEvents = (mode) => {
    // do some mode validation
    return async () => {
        // mode-specific logic to create stream boundary for request
        // perform request and return response
    };
}
```

And then we would have to make 3 separate `useQuery` calls anyway (using `getEvents("default")` or `getEvents("newer")` or `getEvents("older")` as the `queryFn` for each call, each with pretty different `onSuccess` handlers.

Removing the query-specific logic from where we define the queries and moving it to `getEvents` makes things a lot more confusing when reading the code.

@peersky: Please do let me know if there's a better way. But let's save it for a separate PR.

Reply:
@zomglings I think refetch here won't work, my mistake. What would work is to define fetch function with an argument. 

Here is example how it's done in bugout search: 


```js
const useJournalEntries = ({
  journalId,
  journalType,
  isContent,
  pageSize,
  searchQuery,
}) => {
  const limit = pageSize ? pageSize : 25;

  const { entriesSearch } = useEntriesSearch({
    journalId,
  });

  const getEntries = (searchTerm) => async (key, jorunalId, nextPage = 0) => {
    if (!nextPage) {
      nextPage = 0;
    }

    const searchTags = searchTerm.split(" ").filter(function (n) {
      if (n.startsWith("#")) return n;
      else {
        return null;
      }
    });

    const data = await entriesSearch({
      searchTerm,
      journalType,
      isContent,
      limit,
      offset: nextPage,
      searchTags,
    });

    const newEntryList = data.data.results.map((entry) => ({
      ...entry,
      id: entry.entry_url.split("/").pop(),
    }));
    return {
      data: [...newEntryList],
      nextPage: nextPage + 1,
      next_offset: data.data.next_offset,
      total_results: data.data.total_results,
      offset: data.data.offset,
    };
  };

  const {
    data: EntriesPages,
    isFetchingMore,
    isLoading,
    canFetchMore,
    fetchMore,
    refetch,
  } = useInfiniteQuery(
    ["journal-entries", { journalId }],
    getEntries(searchQuery),
    {
      ...queryCacheProps,
      getNextPageParam: (lastPage) => lastPage.next_offset ?? false,
      getFetchMore: (lastGroup) => {
        return lastGroup.next_offset === null ? false : lastGroup.next_offset;
      },
      enabled: !!journalId,
    }
  );
```

Here `searchTerm` is being passed as argument.

Implementation for moonsteam would look like 

```js
// /component code 
const [fetchMethod, setFetchMethod] = useState('newest')
const streamCache  = useStream(otherArguments, {fetchMethod})

// hook code 
const useJournalEntries = ({
 fetchMethod,
  journalId,
  journalType,
  isContent,
  pageSize,
  searchQuery,
}) => {

...
useQuery(
    ["stream-entries", queryKeys],
    getStream(fetchMethod),
    .... )

```

In this implementation actually I don't see any problem to store fetchMethod inside queryKeys with name `lastFetchMethod: <Method>` , by that way avoiding need in both function wrapping and extra query hooks.

Reply:
@peersky : I think that's the same as the final option from my comment above. I tried it and in the `useStream` case it really doesn't work well because the different `onSuccess` implementations still make it difficult to work with a single query.

See the order in which each `onSuccess` adds events to the `events` state.

Reply:
Using `switch` in `onSuccess` would not be reliable as it's async. I agree that in this case its lgtm to use multiple queries. 
However it still seems quite confusing. I made an issue to follow this topic: 
https://github.com/bugout-dev/moonstream/issues/158